src/goto-programs/goto_contractor.cpp

```diff
@@ -0,0 +1,453 @@
+//
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#include "goto_contractor.h"
+
+void goto_contractort::get_constraints(goto_functionst goto_functions)
+{
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assert())
+      constraint = create_constraint_from_expr2t(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::get_intervals(goto_functionst goto_functions)
+{
+  //only main here
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assume())
+    {
+      it->dump();
+      parse_intervals(it->guard);
+    }
+
+    it++;
+  }
+}
+
+void goto_contractort::parse_intervals(irep_container<expr2t> expr)
+{
+  symbol2tc symbol;
+  long value;
+
+  expr = get_base_object(expr);
+
+  if(is_and2t(expr))
+  {
+    parse_intervals(to_and2t(expr).side_1);
+    parse_intervals(to_and2t(expr).side_2);
+    return;
+  }
+
+  if(!is_comp_expr(expr))
+    return;
+
+  std::shared_ptr<relation_data> rel;
+  rel = std::dynamic_pointer_cast<relation_data>(expr);
+
+  //side1 1 is a symbol or typecast to symbol
+  auto side1 = rel->side_1;
+  //side 2 is always a number.
+  auto side2 = rel->side_2;
+
+  auto obj = get_base_object(side1);
+  if(is_symbol2t(obj))
```

Comment:
Can you do `if(!is_symbol(obj)) return;`?

Reply:
done.